design/baremetal-operator/firmware-image-update.md

```diff
@@ -0,0 +1,195 @@
+- component – The hardware component which this firmware matches. Hardware vendor may have several types of hardware for example Nvidia has NiC’s and GPU. Current proposal will be only `net` and can be extended in future proposals for other vendors.
+- componentFlavor- indicate the component flavor. Optional attribute. For Nvidia NICs it will be the PSID.
+- version - the requested firmware version of the firmware image.
+
+example of the HostFirmwareImage spec section:
+
+```yaml
+spec:
+  firmwareImages:
+    - url: http://firmware_images/fw1.bin
+      checksum: a94e683ea16d9ae44768f0a65942234d
+      checksumType: md5
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - url: http://firmware_images/fw2.bin
+      checksum: e07da4de09e9a2ab836bd7e23378f8e45afd5400f0f1cf7da5ecefce45333d5d
+      checksumType: sha1
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_0000000652
+      version: 22.35.1012
+```
+
+The `HostFirmwareImage` status section will introduce the following attributes:
+
+- Firmwares – list of all hardware devices with the following information:
+    - componentID – unique identifier for the hardware. For NICs the unique identifier will be in the following format “`<interface name>_<MAC>”.`
+    - vendor – The hardware vendor which is responsible for that firmware.
+    - component – The hardware component type.
+    - componentFlavor - indicate the component flavor if such exist.
+    - version - the actual firmware version of the hardware.
+- lastUpdated – The timestamp of the last update.
+- Conditions – we introduce 2 conditions:
+    - FirmwareImageChangeDetected - Indicates that the firmware images version in the spec are different than status.
+    - FirmwareImageValid - Indicates that firmware images are valid and can be configured on the host.
+
+```yaml
+status:
+  firmwares:
+    - componentID: enp216s0f0np0_08_c0_eb_70_74_62
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - componentID: enp216s0f1np1_08_c0_eb_70_74_63
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_1090120019
+      version: 24.34.1002
+    - componentID: enp59s0f0np0_b8_ce_f6_8d_c8_62
+      vendor: nvidia
+      component: net
+      componentFlavor: MT_0000000652
+      version: 22.35.1012
+  lastUpdated: 2022-12-27T01:12:17Z
+  conditions: ...
+```
+
+### HostFirmwareSettings Controller Changes
+
+The proposal is to extend the `HostFirmwareSettings` controller which was introduced in the `bulk-set-bios-config` proposal. The controller will be extended to track the `HostFirmwareImage` CRD. The HostFirmwareImage.spec is the desired firmware version and the HostFirmwareImage.status contain the actual firmware. Please note that the correct proposal addresses Nvidia NICs but can be extended to other vendors too as well.

```

Comment:
```suggestion
The proposal is to extend the `HostFirmwareSettings` controller which was introduced in the `bulk-set-bios-config` proposal. The controller will be extended to track the `HostFirmwareImage` CRD. The HostFirmwareImage.spec is the desired firmware version and the HostFirmwareImage.status contain the actual firmware. Please note that the current proposal addresses Nvidia NICs but can be extended to other vendors too as well.

```

FWIW, each CRD generally has its own controller, and they all run in the same process. So we wouldn't extend the existing controller, but rather create a new one in the same vein.